enhancements/proposals/cleanup_after_scaling.md

```diff
@@ -0,0 +1,153 @@
+
+## Design Details
+
+Scylla-operator will be extended with a controller responsible for running a cleanup procedure when needed.
+
+### Cleanup trigger
+
+The root cause of the cleanup requirement is that during cluster changes tokens are moved between nodes.
+There are two alternatives of detecting when this happened:
+
+1. ScyllaCluster topology changes - new rack is added/removed, number of members changes, or additional CPUs are
+   added/removed.
+2. Check tokens in Scylla API, detect when it changed.
+
+The first one is edge driven, there are multiple cases when tokens might move or not, for example if cpu resource change
+won’t be big enough to allocate additional core the cleanup is not needed. For hybrid clusters or multi datacenter
+clusters, topology change in the non-managed part should cause the cleanup of nodes in the managed part which is
+impossible to detect. It’s also harder to implement since StatefulSets, which changes need to be detected, doesn't allow
+for per node distinction, the state needs to be carried to other resources (for example member Service) which brings
+unnecessary coupling.
+
+The second one is checking the actual root cause, and seems more bulletproof.
+
+### Implementation
+
+Two new annotations in member services will be reconciled. Controller in the sidecar will
+reconcile `internal.scylla-operator.scylladb.com/current-token-ring-hash`
+and Operator will reconcile `internal.scylla-operator.scylladb.com/last-cleaned-up-token-ring-hash`.
+
+The `current-token-ring-hash` one will always reflect the hashed output of Scylla API `/storage_service/tokens` endpoint
+from a given
+node. The `last-cleaned-up-token-ring-hash` will be initially empty and set after a cleanup job finishes
+to `current-token-ring-hash` value that was present at the time the cleanup job was created.
+
+The output of tokens endpoint is json formatted sorted array of strings (representing numbers), the annotation will hold
+SHA512
+hash of it as a value.
+
+The cleanup call is synchronous, client only triggers the operation, and all the resource intensive operations are done
+by Scylla. The duration of the operation is tied to the size of the entire dataset and disk performance. Cleanup on
+nodes
+that don't require it, also impacts the disk IO.
+To prevent the controller sync loop from being too long, the cleanup controller in Operator will be responsible for
+managing cleanup Jobs. The preconditions for Job for given node are:
+
+* `current-token-ring-hash` and `last-cleaned-up-token-ring-hash` token ring hash annotations values are different.
+* Observed generation is equal to ScyllaCluster generation.
+* Topology changes are done - ScyllaCluster Status contains Condition `StatefulSetControllerProgressing=False`.
+* Upgrade is not running - Upgrade status in ScyllaCluster status is not set.
+* All nodes are UN - rack status ReadyMembers is equal to member count in ScyllaCluster rack spec.
```

Comment:
can we phrase this as checking `Degraded` and `Available` conditions that already aggregate node statuses?